test/Microsoft.AspNetCore.Server.Kestrel.FunctionalTests/AddressRegistrationTests.cs

```diff
@@ -531,6 +534,37 @@ private static int GetNextPort()
             }
         }
 
+        private static bool CanBindAndConnectToEndpoint(IPEndPoint endPoint)
+        {
+            try
+            {
+                using (var serverSocket = new Socket(endPoint.AddressFamily, SocketType.Stream, ProtocolType.Tcp))
+                {
+                    serverSocket.Bind(endPoint);
+                    serverSocket.Listen(1);
+
+                    var socketArgs = new SocketAsyncEventArgs
+                    {
+                        RemoteEndPoint = serverSocket.LocalEndPoint
+                    };
+
+                    var mre = new ManualResetEventSlim();
+                    socketArgs.Completed += (s, e) =>
+                    {
+                        mre.Set();
+                        e.ConnectSocket.Dispose();
+                    };
+
+                    // Connect can take a couple minutes to time out.
+                    return Socket.ConnectAsync(SocketType.Stream, ProtocolType.Tcp, socketArgs) && mre.Wait(5000);
```

Comment:
`ConnectAsync` will return false if the operation completes synchronously. In that case the callback is not called and you have to call it manually. You have to check the result in the callback to make sure the connect really succeeded.